// tmdb/dist/errors.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.RemoteError = exports.NotFoundError = exports.UnexpectedResponseError = exports.Unimplemented = exports.TmdbError = void 0;

var _es6Error = _interopRequireDefault(require("es6-error"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

class TmdbError extends _es6Error.default {}exports.TmdbError = TmdbError;



class Unimplemented extends TmdbError {
  constructor() {
    super('Method/ behaviour is not implemented.');
  }}exports.Unimplemented = Unimplemented;


class UnexpectedResponseError extends TmdbError {
  constructor() {
    super('Remote service produced an unexpected response.');
  }}exports.UnexpectedResponseError = UnexpectedResponseError;


class NotFoundError extends TmdbError {
  constructor() {
    super('Resource not found.');
  }}exports.NotFoundError = NotFoundError;


class RemoteError extends TmdbError {




  constructor(message /*: string*/, code /*: number*/) {
    super();

    this.code = code;
    this.message = message;
  }}exports.RemoteError = RemoteError;
